### ISSUE TITLE
Make log an optional dependency behind "logging"

### ISSUE THREAD

**Issue by EriKWDev** [rust-windowing/winit#2541]
log could be enabled when feature "logging" is enabled and it could be included by default.

We can define a log! (info!, warn!...) macro with same signature as that from the log crate which calls log's macros if we have the feature enabled and otherwise do a no-op.

Seems like a low-hanging fruit to get one dependency less and It really shouldn't be required by a windowing library I feel.

**Comment by kchibisov** [comment]
The log is actually very important if you want to get any help on macOS or if you want to know why something is not working. So I'm against that, since the information log is telling is helpful for debugging purposes and for reporting issues.

What's your issue with `log` in particular?

**Comment by EriKWDev** [comment]
I don't have an issue with log in particular, just don't like when crates give me transient dependencies on crates that aren't necessary. 

I feel like log isn't a critical dependency for a windowing library, so then I should be able to turn it off. I still purpose it to be enabled by default.

**Comment by necauqua** [comment]
`log` is the farthest thing from "bloat" there could be, gave me a laugh - it's purpose (one of them) is literally to prevent bloat by decoupling the facade from implementations.

It is a ubiquitous (and tiny) facade which on it's own does nothing, depends on nothing (except cfg_if, which is a _macro_ library just for itself, lol) and requires an implementation part to be added and enabled _by you_ in application code. `simple_logger` you see is a dev-dependency for examples/tests.

The `log` in and of itself does not "bloat" the resulting executable/library more than absolutely necessary - well because thankfully Rust is not JS, and cargo took mostly the positives from npm.

Almost every even remotely mature library depends on it anyway - and your code would depend on it too if you don't want to use println! for logging;
Well there's a whole separate ecosystem with `tracing` - for async stuff, but I think it has compatibility with `log` (well, libs that use it) still.

<details>
<summary>cringe rant nobody asked for but you (reader) will still unfold this because you are human</summary>
<br>

> I feel like log isn't a critical dependency for a windowing library

You are wrong.
End-user logging from a windowing library that does the heavy lifting of abstracting over windowing on all major OSes and has to deal with copius amounts of their quirks is NOT "unnecessary", and yes - very much critical.

> a transient dependency for no reason in libraries

You are delusional - well I've described why this sentence is incorrect in previous paragraphs of my ~~useless long internet essay~~ message

> cargo tree | grep log

So you've seen THE SAME CRATE being depended on by 5 libraries and one implementation and you're like "this is so much bloat, much wow" - like what?.

Ok the more I write the more angry at a random person on the internet having a (subjectively) ~~cringe~~ bad take for no reason at all.. If anyone has read all of this through sorry for wasting your time :shrug: 

Ugh I'm back - you are blatantly incorrect even in the "removal" part - watch this trick: https://docs.rs/log/latest/log/#compile-time-filters
If you want log to not leave a trace in the resulting binary.

this guy: "oh yeah lets not have debug info in our rust programs, it takes so much space, ugh, and _It really shouldn't be required by a Rust program I feel_"
</details>

**Comment by EriKWDev** [comment]
Cringw it was indeeed. @necauqua, I feel like you are bringing in a lot of emotion into your answer without reading what I have been saying. I stated, as you yourself, that ’log’ specifically isn’t the problem here:

> I don't have an issue with log in particular, just don't like when crates give me transient dependencies on crates that aren't necessary

I propose to be able to turn off dependencies that I don’ want, and given how trivial the crate’s functionality is I don’t feel like I’m asking for much.

I also didn’t say that end-user logging isn’t crucial, I said that the crate soecifically isn’t crucial. But again, to reiterate what I’ve written many times, this is barely a luke-warm take in my opinion: The user of a library should be able to pick and choose what parts of it they deem necessary. I don’t feel like logging is necessary to open a window, so I would like to be able to turn off the dependency.

let’s keep this a bit more level-headed

**Comment by necauqua** [comment]
It seems that you have the general negative optinion towards bloat (which is understandable and I mean who wouldn't agree with that) but you are putting the effort in the wrong direction.

`log` itself is a almost a no-op (unlike the mentioned log4j) - and is based on macros an whatnot and you can erase it completely (or partially) with a feature-flag that I pointed to.

> The general problem is dependong on libraries that are not necessary.

And we all agree! `isEven` is trash. What is not trash is `log`, they are uncomparable.

You're being like "oh, everyone depends on this one little bit, maybe it's not because it's overwhemingly useful and was _designed_ as a bridge to be used by everywhere, no-no-no, it must be the isEven situation"

**Comment by EriKWDev** [comment]
> "the log crate is not necessary" even sounds wrong 

If a library cannot be written without the log crate in your opinion then I feel like we are not on the same page and will have to agree to disagree. But 'log' is not the issue as I've stated multiple times. It is a no-op. I agree. Chill.

Please address the real issues I brought up in my response, not attack me as a person. I value discussion and If I am wrong I will accept that, but I have not been convinced by your arguments yet.

**Comment by necauqua** [comment]
What, in the previous message I did not attack you as a person _at all_, the only slightly questionable part was the analogy where I did the "you're being like" part?.
You're accusing me, huh.

Anyway.
A library with the log crate and a library without the log crate are not equal in terms of capabilities - and you can turn the first into the latter with the feature-flag, the log crate specifically has this functionality.

> 'log' is not the issue as I've stated multiple times

Then how does removing log solve the global issue that we both agree has its presence?
Repeating myself, you're "putting the effort in the wrong direction".
The log crate is the corner-case where "everyone depends on it and uses it" is a good thing.

If you agree with that then we don't have a disagreement, only normal human miscommunication (and I got angry in the first message, sorry for that).

**Comment by EriKWDev** [comment]
The "you're being like" part was the thing I meant yes. Thanks for acknowladging it was a bit questionable.

The 'log' crate provides good functionality if we want a log. I want that functionality sometimes and usually enable it on other crates.

The core thing I want is to be able to enable and disable functionality that is note core to that crate, which cargo fantastically provides a solution to through 'features' which I love.

I still don't think it's a bizarre statement to make that the log crate is unnecessary for a windowing library.

Communication on the internet and in text is hard and I will admit that I was a bit salty in the first comment I made. I will go touch some grass now before responding further.

**Comment by necauqua** [comment]
Yeah "I still don't think it's a bizarre statement" is the narrowest point of disagreement it seems, oh well. I apologize again for the full-blast cringe rant in the first message, and yes text is so hard and imprecise, thanks for being reasonable (unlike a lot of people you'd meet on the internet).

One last thing:
> The core thing I want is to be able to enable and disable functionality that is not core to a crate, which cargo fantastically provides a solution to through 'features' which I love.

The log crate specifically did use 'features' to be able to get self-erased (as I've said a couple of times), given the ubiquitousness of it - and the fact that everyone would have to reimplement basically the same thing otherwise (and it would not be as convenient for all lib authors to use too). So they solved your issue basically IMO

And yep usually good libraries do just that with the 'features' thing.
